lib/Dialect/SV/Transforms/SVExtractTestCode.cpp

```diff
@@ -269,6 +272,72 @@ static void migrateOps(hw::HWModuleOp oldMod, hw::HWModuleOp newMod,
     }
 }
 
+// Check if the module has already been bound.
+static bool isBound(hw::HWModuleLike op, hw::InstanceGraph &instanceGraph) {
+  auto *node = instanceGraph.lookup(op);
+  return llvm::any_of(node->uses(), [](hw::InstanceRecord *a) {
+    auto inst = a->getInstance();
+    if (!inst)
+      return false;
+    return inst->hasAttr("doNotPrint");
+  });
+}
+
+// Move any old modules that are test code only to the test code area.
+static void maybeMoveToTestCode(hw::HWModuleOp oldMod, Attribute testBenchDir,
+                                Attribute bindFile,
+                                hw::InstanceGraph &instanceGraph) {
+  // Ensure we have a valid test code path.
+  if (!testBenchDir)
+    return;
+
+  // Check if the module only has inputs.
+  if (oldMod.getNumOutputs() != 0)
+    return;
+
+  // Check if this module has been bound already, and return early if so. This
+  // can happen, for example, if a module is input only, but exists to feed both
+  // asserts and covers. In such cases, the binds end up in the bind file for
+  // the first kind of test code statement we saw, and we expect this to not
+  // cause issues.
+  if (isBound(oldMod, instanceGraph)) {
+    LLVM_DEBUG(oldMod->emitWarning("already bound, skipping for ") << bindFile);
```

Comment:
This is a good way of handling warnings like this. I am of the opinion that we should not introduce warnings into release builds. This is a slippery slope for hardware tools that somehow always ends up at 100k+ warnings.